dom.go

```diff
@@ -252,19 +268,59 @@ func (h *HTML) reconcile(prev *HTML) []Mounter {
 		h.node.Set("innerHTML", h.innerHTML)
 	}
 
-	return h.reconcileChildren(prev, nil)
+	return h.reconcileChildren(prev)
 }
 
 // reconcileChildren reconciles children of the current HTML against a previous
 // render's DOM nodes.
-func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts []Mounter) {
-	// TODO better list element reuse
+func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
+	var keyed bool
 	for i, nextChild := range h.children {
-		// TODO(pdf): Add tests for node equality
-		new := h.node != prev.node
-		if i >= len(prev.children) || new {
-			if nextChildList, ok := nextChild.(List); ok {
-				nextChildList.reconcile(h.node, insertBefore, nil)
+		// Massage concrete type if necessary.
+		switch v := nextChild.(type) {
+		case *HTML:
+			if v == nil {
+				nextChild = nil
+				h.children[i] = nextChild
+			}
+		case List:
+			nextChild = childList{html: &HTML{children: v}}
+			h.children[i] = nextChild
+		}
+
+		// TODO(pdf): Add tests for node equality, keyed children
+		var (
+			new                   = h.node != prev.node
+			nextKey               string
+			stableKey             bool
+			insertBeforeKeyedNode jsObject
+		)
+		if keyer, ok := nextChild.(Keyer); ok {
+			nextKey = keyer.Key()
+			if len(h.keyMap) > 0 && nextKey == "" {
+				panic("vecty: all siblings must have keys when using keyed elements")
+			}
+			if nextKey != "" {
+				if h.keyMap == nil {
+					h.keyMap = make(map[string]ComponentOrHTML)
+				}
+				if _, exists := h.keyMap[nextKey]; exists {
+					panic("vecty: duplicate sibling key: " + nextKey)
+				}
+				h.keyMap[nextKey] = nextChild
+			}
+		}
+		if !keyed {
+			keyed = len(h.keyMap) > 0
+		}
```

Comment:
Does `keyed` need to be persistent outside the scope of the for loop?

i.e., why don't we just declare `keyed := len(h.keyMap) > 0` before the first usage below here?

Reply:
I think this is fine.

Reply:
By "I think this is fine" do you mean the current code, or what I suggested?

Reply:
I think the change is fine.

Reply:
This does need to exist outside this scope, since we use it to determine how to remove children.  I've slightly changed how this works in the next push any way, since we need to track this across renders too now.